utilities/packages/python/provena-shared-functionality/ProvenaSharedFunctionality/utils.py

```diff
@@ -0,0 +1,104 @@
+from typing import Optional, Dict
+import re
+import os
+
+# This allows users to refer to the value of the resolved variable in conditional statements e.g. 
+# "${VALUE?'!!_postfix'||null}"
+SELF_REFERENCE_MARKER = "!!"
+
+
+def perform_config_substitution(
+    input_string: str, replacements: Optional[Dict[str, str]] = None
+) -> str:
+    """
+    Perform case-insensitive, bash-style variable substitution on the input string with extended functionality.
+
+    Note that in all cases, any single quotes ' will be replaced with double quotes in the output ".
+
+    Please also note the special case of a conditional statement immediately surrounded by quotes. 
+    In this case the double quotes are stripped. But you can reintroduce by using single quotes. This 
+    allows the following style  "${VALUE?'!!'||null}" which remains valid in the JSON document.
+
+    The function supports the following patterns (in order of precedence):
+    1. "${VARIABLE_NAME?true_value||false_value}" - Quoted conditional substitution with else value
+    2. "${VARIABLE_NAME?true_value}" - Quoted conditional substitution without else value
+    3. ${VARIABLE_NAME?true_value||false_value} - Unquoted conditional substitution with else value
+    4. ${VARIABLE_NAME?true_value} - Unquoted conditional substitution without else value
+    5. ${VARIABLE_NAME:-default_value} - Substitution with default value
+    6. ${VARIABLE_NAME} - Basic substitution
+
+    Args:
+        input_string (str): The input string containing variables to be substituted.
+        replacements (Optional[Dict[str, str]]): An optional dictionary of replacement keys -> values
+            to be used if the environment variable isn't available.
+
+    Returns:
+        str: The input string with all variables substituted.
+    """
+
+    def get_replacement(var_name: str) -> Optional[str]:
+        """Get environment variable value case-insensitively."""
+        for key, value in os.environ.items():
+            if key.lower() == var_name.lower():
+                return value
+        if replacements:
+            for key, value in replacements.items():
+                if key.lower() == var_name.lower():
+                    return value
+        return None
+
+    def is_truthy(value: Optional[str]) -> bool:
+        """Determine if a value is truthy."""
+        return value is not None and value.lower() != "false"
+
+    def replace_quotes(s: str) -> str:
+        """Replace single quotes with double quotes, except for 'null'."""
+        return "null" if s.strip() == "null" else s.replace("'", '"')
```

Comment:
What about apostrophes? E.g. Don't -> Don"t. Is that what we want? will input to this function have apostrophes? Need to also check for Null or NULL?

Reply:
The doc string is out of date

Reply:
this is actually incorrect from what I wanted it to do - I need to update this

Reply:
changing to simply return s.replace("'", '"')